atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraTracingConfig.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.cassandra;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.UnsafeArg;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Value.Immutable
+@JsonDeserialize(as = ImmutableCassandraTracingConfig.class)
+@JsonSerialize(as = ImmutableCassandraTracingConfig.class)
+public abstract class CassandraTracingConfig {
+    private static final Logger log = LoggerFactory.getLogger(CassandraTracingConfig.class);
+    private static final AtomicBoolean loggedWarning = new AtomicBoolean();
+
+    @JsonProperty("enabled")
+    @Value.Default
+    public boolean enabled() {
+        return false;
+    }
+
+    @JsonProperty("trace-probability")
+    @Value.Default
+    public double traceProbability() {
+        return 1.0;
+    }
+
+    @JsonProperty("tables-to-trace")
+    public abstract Set<String> tablesToTrace();
+
+    @JsonProperty("min-duration-to-log-ms")
+    @Value.Default
+    public int minDurationToLogMs() {
+        return 0;
+    }
+
+    public boolean shouldTraceQuery(String tablename, Random random) {
+        if (!enabled()) {
+            return false;
+        }
+        if (tablesToTrace().contains(tablename)) {
+            if (traceProbability() == 1.0) {
+                return true;
+            } else {
+                if (random.nextDouble() <= traceProbability()) {
+                    return true;
+                }
+            }
+        }
+        if (tablesToTrace().isEmpty()) {
+            return true; // accept enabled = true but no tables specified to mean trace all tables
+        }
+        return false;
+    }
+
+    @Value.Check
+    protected void check() {
+        // This does not actually validate anything but mimics the old behavior where enabling tracing
+        // would log a warning.
+        if (enabled()) {
+            if (loggedWarning.compareAndSet(false, true)) {
```

Comment:
Do we really need this atomic boolean to mimic the old behaviour? It made sense before because it was periodically reloading the file, but is this object periodically recreated?

Reply:
This code will trigger every time the runtime config is parsed, which by default in witchcraft is every 5 seconds. This will be a ton more spammy than before if we removed this.

Reply:
Interesting. I wasn't aware that it actually creates this object that frequently.

What happens if:
1. there are multiple transaction managers in the same JVM (and this is configured for some but not all)?
2. the tracer is enabled, disabled, then re-enabled?

Reply:
When there are multiple in the same JVM it will only log once.

If enabled, disabled, then re-enabled it still would have only logged the first time.

It will also log an additional time whenever the service restarts.

Reply:
I'm also not aware of any internal products that would create multiple Cassandra KVSes in the same JVM outside of testing.

Reply:
Also looked at the witchcraft code again, it actually defaults to reparse config every 1 second.